File: src/main/java/org/gga/graph/search/bfs/ParentRecorder.java
package org.gga.graph.search.bfs;

import org.gga.graph.Edge;
import org.gga.graph.Graph;

import java.util.Arrays;

/**
 * @author mike
 */
public class ParentRecorder extends AbstractBfsVisitor {
    private final int[] parents;

    public ParentRecorder(int[] parents) {
        this.parents = parents;
        Arrays.fill(parents, -1);
    }

    @Override
    public void treeEdge(Edge e, Graph g) {
        parents[e.w()] = e.v();
    }

    public static int[] computeParents(Graph g, int firstVertex) {
        int[] parents = new int[g.V()];
        BreadthFirstSearch.breadthFirstSearch(g, new ParentRecorder(parents), firstVertex);
        return parents;
    }
}

File: src/main/scala/org/gga/graph/search/bfs/ParentRecorder.scala
package org.gga.graph.search.bfs

import org.gga.graph.Edge
import org.gga.graph.Graph
import java.util.Arrays
import java.util

/**
 * @author mike
 */
object ParentRecorder {
  def computeParents(g: Graph, firstVertex: Int): Array[Int] = {
    val parents = new Array[Int](g.V)
    BreadthFirstSearch.breadthFirstSearch(g, new ParentRecorder(parents), firstVertex)
    parents
  }
}

class ParentRecorder(aParents: Array[Int]) extends AbstractBfsVisitor {
  private final val parents: Array[Int] = aParents

  util.Arrays.fill(parents, -1)

  override def treeEdge(e: Edge, g: Graph) {
    parents(e.w) = e.v
  }

}